File: src/branchnexus/session.py
"""Session shutdown and cleanup handler."""

from __future__ import annotations

import logging as py_logging
import subprocess
from collections.abc import Callable
from dataclasses import dataclass
from enum import Enum
from typing import Any

from branchnexus.terminal.models import RuntimeKind, TerminalInstance
from branchnexus.worktree.manager import ManagedWorktree, WorktreeManager

logger = py_logging.getLogger(__name__)


def _validate_terminal_count(value: int) -> int:
    if value < 2 or value > 16:
        raise ValueError(f"Invalid terminal count: {value}")
    return value


class ExitChoice(str, Enum):
    CANCEL = "Vazgec"
    PRESERVE = "Koruyarak Cik"
    CLEAN = "Temizleyerek Cik"


@dataclass
class SessionCleanupResult:
    closed: bool
    cancelled: bool
    removed: list[str]
    preserved_dirty: list[str]


@dataclass(frozen=True)
class SessionTerminalSnapshot:
    terminal_id: str
    title: str
    runtime: RuntimeKind
    repo_path: str
    branch: str


@dataclass(frozen=True)
class RuntimeSessionSnapshot:
    layout: str
    template_count: int
    focused_terminal_id: str
    terminals: list[SessionTerminalSnapshot]

    def to_dict(self) -> dict[str, object]:
        return {
            "layout": self.layout,
            "template_count": self.template_count,
            "focused_terminal_id": self.focused_terminal_id,
            "terminals": [
                {
                    "terminal_id": item.terminal_id,
                    "title": item.title,
                    "runtime": item.runtime.value,
                    "repo_path": item.repo_path,
                    "branch": item.branch,
                }
                for item in self.terminals
            ],
        }


class SessionCleanupHandler:
    def __init__(self, manager: WorktreeManager, prompt: Callable[[list[str]], ExitChoice]) -> None:
        self.manager = manager
        self.prompt = prompt

    def handle_exit(
        self,
        *,
        runner: Callable[..., subprocess.CompletedProcess],
    ) -> SessionCleanupResult:
        if self.manager.cleanup_policy == "persistent":
            logger.debug("Session cleanup skipped due to persistent policy")
            return SessionCleanupResult(
                closed=True,
                cancelled=False,
                removed=[],
                preserved_dirty=[str(item.path) for item in self.manager.managed],
            )

        dirty: list[ManagedWorktree] = []
        clean: list[ManagedWorktree] = []
        for worktree in self.manager.managed:
            if self.manager.check_dirty(worktree, runner=runner):
                dirty.append(worktree)
            else:
                clean.append(worktree)
        logger.debug("Cleanup check complete dirty=%s clean=%s", len(dirty), len(clean))

        if not dirty:
            removed = self.manager.cleanup(runner=runner)
            logger.debug("All worktrees clean; removed=%s", len(removed))
            return SessionCleanupResult(
                closed=True,
                cancelled=False,
                removed=[str(item) for item in removed],
                preserved_dirty=[],
            )

        choice = self.prompt([str(item.path) for item in dirty])
        logger.info("Cleanup prompt result choice=%s dirty_count=%s", choice, len(dirty))
        if choice == ExitChoice.CANCEL:
            return SessionCleanupResult(
                closed=False,
                cancelled=True,
                removed=[],
                preserved_dirty=[str(item.path) for item in dirty],
            )

        if choice == ExitChoice.PRESERVE:
            removed_clean = self.manager.cleanup(runner=runner, selected=clean)
            logger.debug("Preserve dirty worktrees; removed clean=%s", len(removed_clean))
            return SessionCleanupResult(
                closed=True,
                cancelled=False,
                removed=[str(item) for item in removed_clean],
                preserved_dirty=[str(item.path) for item in dirty],
            )

        removed_all = self.manager.cleanup(runner=runner)
        logger.debug("Cleanup forced for all worktrees removed=%s", len(removed_all))
        return SessionCleanupResult(
            closed=True,
            cancelled=False,
            removed=[str(item) for item in removed_all],
            preserved_dirty=[],
        )


def build_runtime_snapshot(
    *,
    layout: str,
    template_count: int,
    terminals: list[TerminalInstance],
    focused_terminal_id: str = "",
) -> RuntimeSessionSnapshot:
    count = _validate_terminal_count(template_count)
    snapshots = [
        SessionTerminalSnapshot(
            terminal_id=item.spec.terminal_id,
            title=item.spec.title,
            runtime=item.spec.runtime,
            repo_path=item.spec.repo_path,
            branch=item.spec.branch,
        )
        for item in terminals
    ]
    return RuntimeSessionSnapshot(
        layout=layout,
        template_count=count,
        focused_terminal_id=focused_terminal_id,
        terminals=snapshots,
    )


def parse_runtime_snapshot(raw: Any) -> RuntimeSessionSnapshot | None:
    if not isinstance(raw, dict):
        return None
    terminals_raw = raw.get("terminals")
    if not isinstance(terminals_raw, list):
        return None
    if not terminals_raw:
        return None

    try:
        template_count = _validate_terminal_count(int(raw.get("template_count", len(terminals_raw))))
    except Exception:
        return None

    terminals: list[SessionTerminalSnapshot] = []
    for item in terminals_raw:
        if not isinstance(item, dict):
            return None
        terminal_id = str(item.get("terminal_id", "")).strip()
        title = str(item.get("title", terminal_id or "Terminal")).strip()
        repo_path = str(item.get("repo_path", "")).strip()
        branch = str(item.get("branch", "")).strip()
        runtime_raw = str(item.get("runtime", RuntimeKind.WSL.value)).strip().lower()
        if runtime_raw == RuntimeKind.POWERSHELL.value:
            runtime = RuntimeKind.POWERSHELL
        elif runtime_raw == RuntimeKind.WSL.value:
            runtime = RuntimeKind.WSL
        else:
            return None
        if not terminal_id:
            return None
        terminals.append(
            SessionTerminalSnapshot(
                terminal_id=terminal_id,
                title=title,
                runtime=runtime,
                repo_path=repo_path,
                branch=branch,
            )
        )

    return RuntimeSessionSnapshot(
        layout=str(raw.get("layout", "grid")),
        template_count=template_count,
        focused_terminal_id=str(raw.get("focused_terminal_id", "")).strip(),
        terminals=terminals,
    )

File: ts-src/core/session.ts
import {
  type RuntimeSessionSnapshot,
  type SessionTerminalSnapshot,
  type SessionCleanupResult,
  type ManagedWorktree,
  ExitChoice,
  type RuntimeKind,
  isSessionSnapshot,
} from '../types/index.js';
import { WorktreeManager } from '../git/worktree.js';
import { logger } from '../utils/logger.js';

const TERMINAL_MIN = 2;
const TERMINAL_MAX = 16;

function validateTerminalCount(value: number): number {
  if (value < TERMINAL_MIN || value > TERMINAL_MAX) {
    throw new Error(`Invalid terminal count: ${value}`);
  }
  return value;
}

export function buildRuntimeSnapshot(
  layout: string,
  templateCount: number,
  terminals: SessionTerminalSnapshot[],
  focusedTerminalId = ''
): RuntimeSessionSnapshot {
  const count = validateTerminalCount(templateCount);
  return {
    layout,
    templateCount: count,
    focusedTerminalId,
    terminals,
  };
}

export function parseRuntimeSnapshot(raw: unknown): RuntimeSessionSnapshot | null {
  if (!isSessionSnapshot(raw)) {
    return null;
  }

  const snapshot = raw;

  if (snapshot.terminals.length === 0) {
    return null;
  }

  try {
    validateTerminalCount(snapshot.templateCount);
  } catch {
    return null;
  }

  const terminals: SessionTerminalSnapshot[] = [];
  for (const item of snapshot.terminals) {
    if (!item.terminalId) {
      return null;
    }

    const runtime = normalizeRuntimeKind(item.runtime);
    if (!runtime) {
      return null;
    }

    terminals.push({
      terminalId: item.terminalId.trim(),
      title: item.title?.trim() || item.terminalId,
      runtime,
      repoPath: item.repoPath?.trim() || '',
      branch: item.branch?.trim() || '',
    });
  }

  return {
    layout: snapshot.layout || 'grid',
    templateCount: snapshot.templateCount,
    focusedTerminalId: snapshot.focusedTerminalId?.trim() || '',
    terminals,
  };
}

function normalizeRuntimeKind(value: string): RuntimeKind | null {
  const normalized = value?.toLowerCase().trim();
  if (normalized === 'wsl') return 'wsl';
  if (normalized === 'powershell') return 'powershell';
  if (normalized === 'native') return 'native';
  return null;
}

export class SessionCleanupHandler {
  private manager: WorktreeManager;
  private prompt: (dirty: string[]) => Promise<ExitChoice>;
  private distribution?: string;

  constructor(
    manager: WorktreeManager,
    prompt: (dirty: string[]) => Promise<ExitChoice>,
    distribution?: string
  ) {
    this.manager = manager;
    this.prompt = prompt;
    this.distribution = distribution;
  }

  async handleExit(): Promise<SessionCleanupResult> {
    if (this.manager.getCleanupPolicy() === 'persistent') {
      logger.debug('Session cleanup skipped due to persistent policy');
      return {
        closed: true,
        cancelled: false,
        removed: [],
        preservedDirty: this.manager.getManaged().map((w) => w.path),
      };
    }

    const dirty: ManagedWorktree[] = [];
    const clean: ManagedWorktree[] = [];

    for (const worktree of this.manager.getManaged()) {
      const isDirty = await this.manager.checkDirty(worktree, this.distribution);
      if (isDirty) {
        dirty.push(worktree);
      } else {
        clean.push(worktree);
      }
    }

    logger.debug(`Cleanup check complete dirty=${dirty.length} clean=${clean.length}`);

    if (dirty.length === 0) {
      const removed = await this.manager.cleanup({
        distribution: this.distribution,
      });
      logger.debug(`All worktrees clean; removed=${removed.length}`);
      return {
        closed: true,
        cancelled: false,
        removed,
        preservedDirty: [],
      };
    }

    const choice = await this.prompt(dirty.map((w) => w.path));
    logger.info(`Cleanup prompt result choice=${choice} dirty_count=${dirty.length}`);

    if (choice === ExitChoice.CANCEL) {
      return {
        closed: false,
        cancelled: true,
        removed: [],
        preservedDirty: dirty.map((w) => w.path),
      };
    }

    if (choice === ExitChoice.PRESERVE) {
      const removedClean = await this.manager.cleanup({
        selected: clean,
        distribution: this.distribution,
      });
      logger.debug(`Preserve dirty worktrees; removed clean=${removedClean.length}`);
      return {
        closed: true,
        cancelled: false,
        removed: removedClean,
        preservedDirty: dirty.map((w) => w.path),
      };
    }

    const removedAll = await this.manager.cleanup({
      distribution: this.distribution,
    });
    logger.debug(`Cleanup forced for all worktrees removed=${removedAll.length}`);
    return {
      closed: true,
      cancelled: false,
      removed: removedAll,
      preservedDirty: [],
    };
  }
}
